L1Trigger/L1TTrackMatch/plugins/L1TrackJetClustering.h

```diff
@@ -113,6 +113,77 @@ namespace l1ttrackjet {
     return PassQuality;
   }
 
+  // Eta binning following the hardware logic
+  inline unsigned int eta_bin_firmwareStyle(int eta_word) {
+    //Function that reads in 15-bit eta word (in two's complement) and returns the index of eta bin in which it belongs
+    //Logic follows exactly according to the firmware
+    //We will first determine if eta is pos/neg from the first bit. Each half of the grid is then split into four coarse bins. Bits 2&3 determine which coarse bin to assign
+    //The coarse bins are split into 5 fine bins, final 13 bits determine which of these coarse bins this track needs to assign
+
+    int eta_coarse = 0;
+    int eta_fine = 0;
+
+    if (eta_word & (1 << 15)) {  //If eta is negative (first/leftmost bit is 1)
+      //Second and third bits contain information about which of four coarse bins
+      eta_coarse = 5 * ((eta_word & (3 << 13)) >> 13);
+    } else {  //else eta is positive (first/leftmost bit is 0)
+      eta_coarse = 5 * (4 + ((eta_word & (3 << 13)) >> 13));
+    }
+
+    //Now get the fine bin index. The numbers correspond to the decimal representation of fine bin edges in binary
+    int j = eta_word & 8191;
+    if (j < 1638)
+      eta_fine = 0;
+    else if (j < 3276)
+      eta_fine = 1;
+    else if (j < 4914)
+      eta_fine = 2;
+    else if (j < 6552)
+      eta_fine = 3;
+    else
+      eta_fine = 4;
+
+    //Final eta bin is coarse bin index + fine bin index, subtract 8 to make eta_bin at eta=-2.4 have index=0
+    int eta_ = (eta_coarse + eta_fine) - 8;
+    return eta_;
+  }
```

Comment:
We agree that these "magic numbers" could be cleaned up to avoid confusion down the line. What is the best way to handle this? Is it best to define each as a constant earlier in the file with clear comment? Or is there a preferred way to do this?

Reply:
@SClarkPhysics Since these are known at compile time, please make each of these a `static constexpr` inside the namespace itself. I think strictly that variables like that are supposed to start with a `k` in CMSSW rules, but I don't think anyone is going to check that hard...

Reply:
See this commit: https://github.com/cms-l1t-offline/cmssw/pull/1170/commits/52d3f4c5bc9278975c7897ac12768950a6a93877

We have tried to define all of the constants as static constexpr within the namespace, as suggested.

Reply:
Perfect, thanks.